File: probs/6.cpp
#include <string>;
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) {
            return s;
        }
        
        char temp[numRows][1001];
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < 1001; j++) {
                temp[i][j] = ' ';
            }
        }

        int acc_x = 1;
        int acc_y = 0;

        int i = 0;
        int j = 0;
        int idx = 0;

        while (idx < s.size()) {
            temp[i][j] = s[idx++];

            if (i == numRows - 1) {
                acc_x = -1;
                acc_y = 1;
            } 
            
            if (i == 0) {
                acc_x = 1;
                acc_y = 0;
            }

            i += acc_x;
            j += acc_y;
        }

        string res;

        for (int a = 0; a < numRows; a++) {
            for (int b = 0; b <= j; b++) {
                if (temp[a][b] != ' ') {
                    res += temp[a][b];
                }
            }
        }

        return res;
    }
};

File: zig/6.zig
const std = @import("std");

fn solution(alloc: std.mem.Allocator, str: []const u8, numRows: usize) ![]const u8 {
    var arr = std.ArrayList([]u8).init(alloc);
    defer {
        for (arr.items) |word| {
            alloc.free(word);
        }
        arr.deinit();
    }

    try arr.resize(numRows);

    for (arr.items) |*row| {
        row.* = try alloc.alloc(u8, str.len);
        @memset(row.*, ' ');
    }

    var flag: bool = true;
    var idx: usize = 0;

    var a: usize = 0;
    var b: usize = 0;

    while (idx < str.len) {
        arr.items[a][b] = str[idx];

        if (a == numRows - 1) {
            flag = false;
        }

        if (a == 0) {
            flag = true;
        }

        if (flag) {
            a += 1;
        } else {
            a -= 1;
            b += 1;
        }

        idx += 1;
    }

    var temp = std.ArrayList(u8).init(alloc);
    defer temp.deinit();

    for (arr.items) |*row| {
        for (row.*) |el| {
            if (el != ' ') {
                try temp.append(el);
            }
        }
    }

    return temp.toOwnedSlice();
}

const expect = std.testing.expect;

test "test case #1" {
    const str = "PAYPALISHIRING";
    const result = try solution(std.testing.allocator, str, 3);
    defer std.testing.allocator.free(result);

    try expect(std.mem.eql(u8, result, "PAHNAPLSIIGYIR"));
}

test "test case #2" {
    const str = "PAYPALISHIRING";
    const result = try solution(std.testing.allocator, str, 4);
    defer std.testing.allocator.free(result);

    try expect(std.mem.eql(u8, result, "PINALSIGYAHRPI"));
}

test "test case #3" {
    const str = "A";
    const result = try solution(std.testing.allocator, str, 1);
    defer std.testing.allocator.free(result);

    try expect(std.mem.eql(u8, result, "A"));
}
